### ISSUE TITLE
About watch interface redefine

### ISSUE THREAD

**Issue by Hubery2757** [coreos/jetcd#154]
I test the latest code, watch callback have been deprecated, now we need use watcher.listen to retrieve the watch event. But I need to get watch event by passive way as the same as watch callback . so is there such mechanism for the latest code?

**Comment by fanminshi** [comment]
@Hubery2757 I wonder why do you want to have a passive and asynchronous way of getting watch event? Why wouldn't ` watcher.listen` do the job for you? the reason to remove the callback is to make watch simpler to use and easier to reason about.

**Comment by heyitsanthony** [comment]
A callback mechanism can be built on top of what's there now. Why does the client need to implement it?

**Comment by xiang90** [comment]
we might add an async helper in the future if there are more and more people wanting it. but for now, we are not going to do it.

**Comment by jalenyang** [comment]
@fanminshi I think the passive event notification event it is better and it is needed to many scenarios.
`watcher.listen` will block the main thread. Anyway like @heyitsanthony  said we can build a callback mechanism on top `watcher.listen`.

**Comment by fanminshi** [comment]
@jalenyang can you provide few scenarios that pass even would help?

**Comment by jalenyang** [comment]
@fanminshi one simple example is the auto service discovery.
I.E.  application A watch with the prefix `app/services`, if there is one new service registered(e.g. `app/services/Service B`) and the application A has some dependencies on Service B, then application B should be able to receive an asynchronous notification and start the business.

**Comment by fanminshi** [comment]
@jalenyang I understand the use case.  However, the question I am having is why is  ` asynchronous notification` prefer over `synchronous` one? Wouldn't `synchronous notification` achieves what you want as well?

**Comment by jalenyang** [comment]
@fanminshi well, one example is when the application startup it will watch the configuration changes.
we don't want the main thread was blocked by the `synchronous ` watch. Also, we need to persistent monitoring the changes of the configuration in etcd.  That means we need to implement the `asynchronous ` watch ourselves with `watcher.listen`.  Surely, we can implement it base on your current design, but why did remove it while the `asynchronous `  watcher had already been implemented in the previous version? One more thing, the API of the etcd endpoint is `asynchronous `, so why the client doesn't support
`asynchronous `  API?  It doesn't make sense.

**Comment by lburgazzoli** [comment]
Although it is is possible to achieve an async notification by wrapping the synchronous one, I think every functionality should be asynchronous first.

Now briefly looking at the code I see that as today a call to listen() submit a new task to a single threaded  executor which creates a Future we blocks on, then one eventually needs to create an additional thread/executor to make watch asynchronous again which seems to me a little expensive.

**Comment by fanminshi** [comment]
`why did remove it while the asynchronous watcher had already been implemented in the previous version?`

The previous `asynchronous` version doesn't have correct behavior as the etcd go client and its callback definition was confusing; it has something called `onResuming` which is hard to reason about what to do with it. I rewrote the watch api to match the synchronous nature of the go client which I think is much simpler to reason about.  

The `asynchronous` watch api seems reasonable. I'll add it into future jetcd release.

**Comment by f4n** [comment]
+1 for an asynchronous watch api

**Comment by RutledgePaulV** [comment]
I was also thrown off by this. Either an asynchronous API or an example of turning the synchronous one into an asynchronous one (without spinning the CPU against the thread or polling?) would be great.
